src/openms/source/FORMAT/HANDLERS/MzMLHandler.cpp

```diff
@@ -3907,88 +3919,252 @@ namespace OpenMS::Internal
         for (Size s_idx = 0; s_idx < exp.size(); ++s_idx)
         {
-          if (!exp[s_idx].getNativeID().has('='))
-          {
-            renew_native_ids = true;
-            break;
-          }
+            if (!exp[s_idx].getNativeID().has('='))
+            {
+                renew_native_ids = true;
+                break;
+            }
         }
-
-        // issue warning if something is wrong
         if (renew_native_ids)
         {
-          warning(STORE, String("Invalid native IDs detected. Using spectrum identifier nativeID format (spectrum=xsd:nonNegativeInteger) for all spectra."));
+            warning(STORE, "Invalid native IDs detected. Using spectrum identifier nativeID format for all spectra.");
         }
-
-        // write actual data
+  
         for (Size s_idx = 0; s_idx < exp.size(); ++s_idx)
         {
-          logger_.setProgress(progress++);
-          const SpectrumType& spec = exp[s_idx];
-          writeSpectrum_(os, spec, s_idx, validator, renew_native_ids, dps);
-          ++stored_spectra;
-        }
-        os << "\t\t</spectrumList>\n";
-      }
+            logger_.setProgress(progress++);
 
-      //--------------------------------------------------------------------------------------------
-      // chromatograms
-      //--------------------------------------------------------------------------------------------
-      if (!exp.getChromatograms().empty())
-      {
-        // INFO : do not try to be smart and skip empty spectra or
-        // chromatograms. There can be very good reasons for this (e.g. if the
-        // meta information needs to be stored here but the actual data is
-        // stored somewhere else).
-        os << "\t\t<chromatogramList count=\"" << exp.getChromatograms().size() << "\" defaultDataProcessingRef=\"dp_sp_0\">\n";
-        for (Size c_idx = 0; c_idx != exp.getChromatograms().size(); ++c_idx)
-        {
-          logger_.setProgress(progress++);
-          const ChromatogramType& chromatogram = exp.getChromatograms()[c_idx];
-          writeChromatogram_(os, chromatogram, c_idx, validator);
-          ++stored_chromatograms;
+            // Calculate offset if indexing is enabled 
+            Int64 offset = -1;
+            if (options_.getWriteIndex())
+            {
+                if (!compress) {
+                    offset = static_cast<Int64>(os.tellp());
+                }
+                else { 
+                    offset = counter_filter.characters();
+                }
+            }
+            std::string native_id = exp[s_idx].getNativeID();
+            if (renew_native_ids)
+            {
+                native_id = "scan=" + String(s_idx);
+            }
+            if (options_.getWriteIndex() && offset != -1)
+            {
+                spectra_offsets_.emplace_back(native_id, offset);
+            }
+            writeSpectrum_(*output_stream, exp[s_idx], s_idx, validator, renew_native_ids, dps);
+            stored_spectra++;
```

Comment:
revert this;
You have all the information you need available inside `writeSpectrum`
(counter_ptr_ and compress_)